### PR TITLE
#26288  Throw an exception when a customer does not exist requests password reset

### LINKED ISSUE #26288: missing Throw in copy pasted code
**Preconditions (*)**
magento 2.3.3

**Steps to reproduce (*)**
go to /Magento/Customer/Model/ForgotPasswordToken/GetCustomerByToken.php of magento installation

**Expected result (*)**
exception should be covered with throw

**Actual result (*)**
Hi, you've got a problem with file https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Customer/Model/ForgotPasswordToken/GetCustomerByToken.php

image

Check this pull request if you wish: https://github.com/magento/magento2/pull/26287

### PR DESCRIPTION
**Pull request by Bal2018**
### Description (*)

When a customer (who does not exist) requests a password reset - Magento will now throw an exception error. 
### Fixed Issues (if relevant)

1. fixes magento/magento2#26288: missing `throw` in copy-pasted code

### Manual testing scenarios (*)
 
1. ...created a customer 
2. ...checked email sent for new customer
3. ...deleted customer
4. ...from email clicked on "reset password"
5. ...when email entered, then  error message thrown

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by lbajsarowicz** [comment]
Nice catch. Let me help you with the description, but also I need you to improve code style (Static Tests)

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Please correct the static tests.
Could you also add Unit Tests for that case, to avoid "losing that" in future*?

* Test coverage is part of Definition of Done.

**Comment by engcom-Echo** [comment]
I will take care of test coverage.

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Framework\Exception\State\ExpiredException;
+use Magento\Framework\Phrase;
+use Magento\Framework\TestFramework\Unit\Helper\ObjectManager as ObjectManagerHelper;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Test GetCustomerByToken class
+ */
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]: This annotation is not necessary :-)
You can also use `@covers` if you really need to express anything as class annotation.

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+class GetCustomerByTokenTest extends TestCase
+{
+    protected const RESET_PASSWORD = 'resetPassword';
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
    private const RESET_PASSWORD = 'resetPassword';
```
Unit Tests are not expected to be inherited, use `private` instead

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    /**
+     * @var SearchCriteriaBuilder|MockObject
+     */
+    protected $searchCriteriaBuilderMock;
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
    private $searchCriteriaBuilderMock;
```

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    /**
+     * @var SearchCriteria|MockObject
+     */
+    protected $searchCriteriaMock;
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
    private $searchCriteriaMock;
```

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    /**
+     * @var CustomerRepositoryInterface|MockObject
+     */
+    protected $customerRepositoryMock;
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
    private $customerRepositoryMock;
```

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    /**
+     * @var CustomerSearchResultsInterface|MockObject
+     */
+    protected $searchResultMock;
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
    private $searchResultMock;
```

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+    /**
+     * @var CustomerInterface|MockObject
+     */
+    protected $customerMock;
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
    private $customerMock;
```

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+    /**
+     * @var GetCustomerByToken;
+     */
+    protected $model;
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
    private $model;
```

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+    /**
+     * @inheritDoc
+     */
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
```
No annotation needed

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    public function setUp()
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
    protected function setUp()
```
`setUp` method is `protected`. There is no reason to change it's visibility (refer: `\PHPUnit\Framework\TestCase::setUp`)

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+            ->willReturn($this->searchCriteriaMock);
+        $this->customerRepositoryMock->expects($this->once())
+            ->method('getList')
+            ->with($this->searchCriteriaMock)
+            ->willReturn($this->searchResultMock);
+    }
+
+    /**
+     * @return void
+     */
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
```
This does not introduce any new information that is already in return type hint.

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    public function testExecute(): void
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]: This is not valid test method name - please be more descriptive with your method names - for example:
`testExecuteReturnsFalseWhenNoItemsAvailable` or `testExecuteIsNotBeingCalledWhenNoItemsAvailable`

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        $totalCount = 1;
+        $this->searchResultMock->expects($this->any())
+            ->method('getTotalCount')
+            ->willReturn($totalCount);
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]: If it does not matter how many the method is being called, this means it's **stub**, not **mock**. Then you should avoid `->expects()` at all.

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+    /**
+     * @return void
+     */
+    public function testExecuteWithNoSuchEntityException(): void
+    {
+        $totalCount = 0;
+        $this->searchResultMock->expects($this->any())
+            ->method('getTotalCount')
+            ->willReturn($totalCount);
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]: If it does not matter how many the method is being called, this means it's stub, not mock. Then you should avoid ->expects() at all.

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+        $this->assertInstanceOf(
+            CustomerInterface::class,
+            $this->model->execute(self::RESET_PASSWORD)
+        );
+    }
+
+    /**
+     * @return void
+     */
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
```
This provides no additional value

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+                ['value' => self::RESET_PASSWORD]
+            )
+        ));
+
+        $this->model->execute(self::RESET_PASSWORD);
+    }
+
+    /**
+     * @return void
+     */
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
```
No additional value provided by docblock

File: app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+    public function testExecuteWithExpireException(): void
+    {
+        $totalCount = 2;
+        $this->searchResultMock->expects($this->any())
+            ->method('getTotalCount')
+            ->willReturn($totalCount);
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Unit/Model/ForgotPasswordToken/GetCustomerByTokenTest.php]:
```suggestion
        $this->searchResultMock->method('getTotalCount')->willReturn($totalCount);
```
If it does not matter how many the method is being called, this means it's stub, not mock. Then you should avoid ->expects() at all.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
@engcom-Echo Please review the comments.

**Review by lbajsarowicz — APPROVED** [review]
Awesome. Thank you!

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7527](https://jira.corp.magento.com/browse/ENGCOM-7527) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Comment by engcom-Echo** [comment]
Failed functional tests not related to the changes in this PR

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
Hi @Bal2018  I mistakenly closed PR.

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-7527](https://jira.corp.magento.com/browse/ENGCOM-7527) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
**Result:**
No exception error is shown
image

**Comment by engcom-Delta** [comment]
**Note:** Functional Tests EE and Integration Tests are failed

**Comment by engcom-Echo** [comment]
@magento run all tests

**Pull request merged**